### ISSUE TITLE
Output Format: Prometheus 

### ISSUE THREAD

**Issue by mrueg** [chubin/wttr.in#446]
I would be interested in a Prometheus-style format output similar to the json output that is already provided. https://prometheus.io/docs/concepts/data_model/

This can be used then to scrape wttr.in easily for monitoring systems.

**Comment by chubin** [comment]
It is an interesting idea.
Maybe we could add additional representation.

Would this work for the beginning?

`jq -r 'keys_unsorted[] as $k | "\($k) \(.[$k])"'`

**Comment by mrueg** [comment]
I don't think that it might work as the RHS needs to be numeric.
https://prometheus.io/docs/practices/naming/
```
temperature_celsius{location="Berlin"} 28.9
```

There's a python client library that probably is the best approach to do the output formatting base on the data object.
https://github.com/prometheus/client_python

**Comment by chubin** [comment]
Do I understand it right, that only current values make sense, and there is no need to export forecast values? At least at the beginning.

Question #2:
if you say that the RHS has to be numeric, do I understand it right, that non-numeric fields should not be exported at all?

**Comment by mrueg** [comment]
Forecast makes sense as well.
Non-numeric fields can be exported as dimensions.

```
  "current_condition": [
    {
      "FeelsLikeC": "21",
      "FeelsLikeF": "70",
      "cloudcover": "0",
      "humidity": "31",
      "localObsDateTime": "2020-05-22 11:44 AM",
      "observation_time": "09:44 AM",
      "precipMM": "0.0",
      "pressure": "1021",
      "temp_C": "21",
      "temp_F": "70",
      "uvIndex": 6,
      "visibility": "10",
      "weatherCode": "113",
      "weatherDesc": [
        {
          "value": "Sunny"
        }
      ],
      "weatherIconUrl": [
        {
          "value": ""
        }
      ],
      "winddir16Point": "SE",
      "winddirDegree": "130",
      "windspeedKmph": "19",
      "windspeedMiles": "12"
    }
  ],
```

could become (i left some of the items out):

```
FeelsLikeC{forecast="0h"} 21
FeelsLikeF{forecast="0h"} 70
cloudcover{forecast="0h"} 0
humidity{forecast="0h"}  31
localObsDateTime_epoch{forecast="0h"} 1590141901
observation_time_hour{forecast="0h"}s 584
weatherDesc{forecast="0h", description="Sunny"} 1
winddir16Point{forecast="0h", direction="SE"} 1
```

for the hourly prediction it could look like this:
```
FeelsLikeC{forecast="1h"} 21
FeelsLikeF{forecast="1h"} 70
cloudcover{forecast="1h"} 0
humidity{forecast="1h"}  31
localObsDateTime_epoch{forecast="1h"} 1590141901
observation_time_hour{forecast="1h"}s 584
weatherDesc{forecast="1h", description="Sunny"} 1
winddir16Point{forecast="1h", direction="SE"} 1
```

**Comment by chubin** [comment]
I've added the first trivial implementation of the Promethus output format.
Could you please take a look at it?

```
$ curl wttr.in/Regensburg?format=p1
FeelsLikeC(forecast="0h") 11
FeelsLikeF(forecast="0h") 53
cloudcover(forecast="0h") 100
humidity(forecast="0h") 97
precipMM(forecast="0h") 6.7
pressure(forecast="0h") 1013
temp_C(forecast="0h") 12
temp_F(forecast="0h") 54
uvIndex(forecast="0h") 1
visibility(forecast="0h") 6
winddirDegree(forecast="0h") 1
windspeedKmph(forecast="0h") 11
windspeedMiles(forecast="0h") 7
weatherDesc{forecast="0h", description="Moderate rain"} 1
winddir16Point{forecast="0h", description="N"} 1
```

What is missing at the moment:

* observation date/time, epoch;
* observation time, today (btw. why is it `_hour` in your example, if units are minutes?)
* forecast
* location name (do we need it? probably yes)
* localized weather condition name (irrelevant for `wttr.in`, aber ist durchaus sinnvoll für de.wttr.in)

If you are going to add it to some monitoring system (and if it will work as expected :),
could you please post a couple of screenshots here, just for better understanding
how it should look like, especially non-numerical fields

**Comment by mrueg** [comment]
Thanks for the initial implementation! 
I've sent in https://github.com/chubin/wttr.in/pull/468

What I would suggest is to make "EXPORTED_FIELDS" a map to comply with prometheus-style metric names: https://prometheus.io/docs/practices/naming/#metric-names 

_hour in my example was a mistake, good catch. It should be minutes.

**Comment by chubin** [comment]
Merged, deployed, please test.

Do you have some nice screenshot with that how it is looking now?

**Comment by mrueg** [comment]
```
curl -s "http://wttr.in/Berlin?format=p1" | promtool check metrics
                                                                                                           
temperature_fahrenheit use base unit "celsius" instead of "fahrenheit"
temperature_feels_like_fahrenheit use base unit "celsius" instead of "fahrenheit"
```

looks good, I'd ignore the complaints around fahrenheit, there might be users needing it.

I don't have any dashboard setup yet, but simply ingested into prometheus.

**Comment by chubin** [comment]
Now one should add the time fields, and the forecast data,
extend the README and one can publish this new feature.
I wonder if anybody would use it, but the idea seems to be very interesting

**Comment by mrueg** [comment]
I added a few more bits here: https://github.com/chubin/wttr.in/pull/471

**Comment by chubin** [comment]
I've cleaned up the code a little bit. I hope you will like it

> Again, not being able to test and probably we should consider using the prometheus python client to create those metrics instead of doing concatenations at some point

Yes sure, we can do it, but I am not sure that it will give us anything,
at least data conversion into the Prometheus format per se
is now implemented literally with one single line

Please test

**Comment by mrueg** [comment]
Thanks for the fix and cleanups, looks good to me! So the last remaining bit is transforming the hourly output, right?

**Comment by chubin** [comment]
Yes, I believe so. It should not bee too hard, because we already have almost everything for that.

What I still can't understand, why do you need Prometheus format (an how only this idea came on your mind?) if you don't even have a Prometheus dashboard?

**Comment by mrueg** [comment]
The next step is to expose it in grafana (that's where I'll create the dashboard in), but to get there I needed the exposed metrics exposed first and ingested in my prometheus instance.

**Comment by chubin** [comment]
Yes, but now it works, i.e. you can already add it to Grafana (presuming you already have it)?
Or do you want to add the hourly forecast first?

**Comment by chubin** [comment]
@mrueg Manuel, no news here?

**Comment by mrueg** [comment]
@chubin sorry, for the late reply. Due to some private events I didn't have time to work on the grafana dashboard yet. For feature parity with the json output the hourly forecast would be perfect.

**Comment by chubin** [comment]
@mrueg Agree. I will add it

**Comment by marijus-ravickas** [comment]
Howdy, amazing functionality. Hope it will move further. Just one notice. 
Would be nice to have town/location label in metric. 
issue: `curl -s 'wttr.in/{Nuremberg,Hamburg,Berlin}?format=p1' | grep -v ^# | sort`
```
 ...
temperature_feels_like_celsius{forecast="current"} -2
temperature_feels_like_celsius{forecast="current"} -4
temperature_feels_like_celsius{forecast="current"} -4
 ...
```
Currently the only way it to create a job in prometheus for each location, as job label will be added automatically by prometheus itself.

**Comment by chubin** [comment]
I returned to this issue and would like to finalize it.

@marijus-ravickas Marjus, it is good idea. Could you give an example how it should look like?
Just a short example, so that I could understand it better

**Comment by marijus-ravickas** [comment]
> @marijus-ravickas Marjus, it is good idea. Could you give an example how it should look like? Just a short example, so that I could understand it better

```shell
curl -s 'wttr.in/{Nuremberg,Hamburg,Berlin}?format=p1' | grep '^windspeed_kmph'                                                                                                                                                                20:56:09
windspeed_kmph{forecast="current"} 13
windspeed_kmph{forecast="current"} 11
windspeed_kmph{forecast="current"} 11
```
Example above, should look like:
```shell
curl -s 'wttr.in/{Nuremberg,Hamburg,Berlin}?format=p1' | grep '^windspeed_kmph'                                                                                                                                                                20:56:09
windspeed_kmph{location="Nuremberg", forecast="current"} 13
windspeed_kmph{location="Hamburg,"forecast="current"} 11
windspeed_kmph{location="Berlin,"forecast="current"} 11
```
this would allow us to use PromQL [functions](https://prometheus.io/docs/prometheus/latest/querying/functions/) with location labels. PromQL example:
```promql
avg by (location) (windspeed_kmph) 
```
I am not an expert in python, but could make a PR for this functionality if You want.

**Comment by marijus-ravickas** [comment]
> @marijus-ravickas Marjus, it is good idea. Could you give an example how it should look like? Just a short example, so that I could understand it better

```shell
curl -s 'wttr.in/{Nuremberg,Hamburg,Berlin}?format=p1' | grep '^windspeed_kmph'                                                                                                                                                                20:56:09
windspeed_kmph{forecast="current"} 13
windspeed_kmph{forecast="current"} 11
windspeed_kmph{forecast="current"} 11
```
Example above, should look like:
```shell
curl -s 'wttr.in/{Nuremberg,Hamburg,Berlin}?format=p1' | grep '^windspeed_kmph'                                                                                                                                                                20:56:09
windspeed_kmph{location="Nuremberg", forecast="current"} 13
windspeed_kmph{location="Hamburg,"forecast="current"} 11
windspeed_kmph{location="Berlin,"forecast="current"} 11
```
this would allow us to use PromQL [functions](https://prometheus.io/docs/prometheus/latest/querying/functions/) with location labels. PromQL example:
```promql
avg by (location) (windspeed_kmph) 
```
I am not an expert in python, but could make a PR for this functionality if You want.

**Comment by marijus-ravickas** [comment]
I could achieve similar results using prometheus static config labels:
```
- job_name: 'wttr_in_baltoji_voke'
  static_configs:
    - targets: ['wttr.in']
      labels:
        location: 'baltoji-voke'
  metrics_path: '/Baltoji-voke'
  params:
    format: ['p1']
- job_name: 'wttr_in_berlin'
  static_configs:
    - targets: ['wttr.in']
      labels:
        location: 'berlin'
  metrics_path: '/Berlin'
  params:
    format: ['p1']
```
This would result in these metrics, as an example:
```
temperature_feels_like_celsius{forecast="current", instance="wttr.in:80", job="wttr_in_baltoji_voke", location="baltoji-voke"}
temperature_feels_like_celsius{forecast="current", instance="wttr.in:80", job="wttr_in_berlin", location="berlin"} 
```

**Comment by cmyrsh** [comment]
Hi @chubin , Great tool.
Any plans to add forecasts on promethus format?

**Comment by chubin** [comment]
@marijus-ravickas 

> I am not an expert in python, but could make a PR for this functionality if You want.

Please go ahead. Just make a PR and we will merge it.

@cmyrsh 

Yes.

Could you please share an example how the response could look like

**Comment by cmyrsh** [comment]
Hi @chubin 

when I run 
 ```curl -X GET "v2.wttr.in/Amsterdam?format=p1"```

the output contains following line

```# HELP windspeed_kmph Wind Speed in Kilometres per Hour
windspeed_kmph{forecast="current"} 22
```

Although it does not contain forecast for wind speed

eg
```# HELP windspeed_kmph Wind Speed in Kilometres per Hour
windspeed_kmph{forecast="1d"} 22
```

That is what I am asking for.

The json output "format=j1" contains this info

**Comment by chubin** [comment]
@cmyrsh now it is clear; we will do it. Should we add forecast for the same time intervals that we have in JSON?

**Comment by Mikle-Bond** [comment]
@chubin Hi. I've noticed that, as you've migrated the project to Golang, there is a regression in prometheus formatter. Approximately since 2026-05-06 all the scrapes fail. I use Victoria Metrics, and from it's output I see that the new service wraps metrics into HTML response.

I assume that's because somewhat broken logic around User-Agents. To reproduce you can try `curl -H 'User-Agent: vm_promscrape'` and the output will have `content-type: text/html; charset=utf-8`.

I would suggest forcing plain-text output for `p1` formatter.

**Comment by chubin** [comment]
Yes, you are right. Thank you for reporting.

**Comment by chubin** [comment]
@Mikle-Bond Please try again

**Comment by Mikle-Bond** [comment]
@chubin Yes, data flows. Thank you for the quick fix!
